src/mumble/GlobalShortcut_win.cpp

```diff
@@ -130,576 +142,395 @@ GlobalShortcutWin::~GlobalShortcutWin() {
-		} else {
-			qWarning(
-				"GlobalShortcutWin: Error looking up LowLevelHooksTimeout. (Error: 0x%lx, Type: 0x%lx, Value: 0x%lx)",
-				static_cast< unsigned long >(err), static_cast< unsigned long >(type),
-				static_cast< unsigned long >(value));
+#ifdef USE_XBOXINPUT
+	if (Global::get().s.bEnableXboxInput) {
+		m_xinput.reset(new XboxInput());
+		if (!m_xinput->isValid()) {
+			qWarning("GlobalShortcutWin: Failed to initialize XboxInput!");
+			m_xinput.reset();
 		}
 	}
-
-	// Wait for MainWindow's constructor to finish before we enumerate DirectInput devices.
-	// We need to do this because adding a new device requires a Window handle. (SetCooperativeLevel())
-	while (!Global::get().mw)
-		this->yieldCurrentThread();
-
+#endif
 #ifdef USE_GKEY
 	if (Global::get().s.bEnableGKey) {
-		gkey = new GKeyLibrary();
-		qWarning("GlobalShortcutWin: GKeys initialized, isValid: %d", gkey->isValid());
+		m_gkey.reset(new GKeyLibrary());
+		if (!m_gkey->isValid()) {
+			qWarning("GlobalShortcutWin: Failed to initialize GKey!");
+			m_gkey.reset();
+		}
 	}
 #endif
-
-#ifdef USE_XBOXINPUT
-	if (Global::get().s.bEnableXboxInput) {
-		xboxinput = new XboxInput();
-		ZeroMemory(&xboxinputLastPacket, sizeof(xboxinputLastPacket));
-		qWarning("GlobalShortcutWin: XboxInput initialized, isValid: %d", xboxinput->isValid());
+	// Wait for MainWindow's constructor to finish, so that we can get the window's handle.
+	while (!mumble_mw_hwnd) {
+		yieldCurrentThread();
 	}
-#endif
 
-	QTimer *timer = new QTimer;
-	connect(timer, SIGNAL(timeout()), this, SLOT(timeTicked()));
-	timer->start(20);
+	RAWINPUTDEVICE rid[5];
 
-	setPriority(QThread::TimeCriticalPriority);
+	rid[0].usUsagePage = HID_USAGE_PAGE_GENERIC;
+	rid[0].usUsage     = HID_USAGE_GENERIC_MOUSE;
+	rid[0].dwFlags     = RIDEV_INPUTSINK;
+	rid[0].hwndTarget  = mumble_mw_hwnd;
 
-	exec();
+	rid[1].usUsagePage = HID_USAGE_PAGE_GENERIC;
+	rid[1].usUsage     = HID_USAGE_GENERIC_JOYSTICK;
+	rid[1].dwFlags     = RIDEV_INPUTSINK | RIDEV_DEVNOTIFY;
+	rid[1].hwndTarget  = mumble_mw_hwnd;
 
-	delete timer;
+	rid[2].usUsagePage = HID_USAGE_PAGE_GENERIC;
+	rid[2].usUsage     = HID_USAGE_GENERIC_GAMEPAD;
+	rid[2].dwFlags     = RIDEV_INPUTSINK | RIDEV_DEVNOTIFY;
+	rid[2].hwndTarget  = mumble_mw_hwnd;
 
-#ifdef USE_GKEY
-	delete gkey;
-#endif
+	rid[3].usUsagePage = HID_USAGE_PAGE_GENERIC;
+	rid[3].usUsage     = HID_USAGE_GENERIC_KEYBOARD;
+	rid[3].dwFlags     = RIDEV_INPUTSINK | RIDEV_DEVNOTIFY;
+	rid[3].hwndTarget  = mumble_mw_hwnd;
 
-#ifdef USE_XBOXINPUT
-	delete xboxinput;
-#endif
+	rid[4].usUsagePage = HID_USAGE_PAGE_GENERIC;
+	rid[4].usUsage     = HID_USAGE_GENERIC_MULTI_AXIS_CONTROLLER;
+	rid[4].dwFlags     = RIDEV_INPUTSINK | RIDEV_DEVNOTIFY;
+	rid[4].hwndTarget  = mumble_mw_hwnd;
 
-	if (bHook) {
-		if (hhMouse) {
-			UnhookWindowsHookEx(hhMouse);
-		}
-		if (hhKeyboard) {
-			UnhookWindowsHookEx(hhKeyboard);
-		}
+	if (!RegisterRawInputDevices(rid, sizeof(rid) / sizeof(rid[0]), sizeof(rid[0]))) {
```

Comment:
If we pack the size of the `rid` array from above into a constexpr variable, we could simply use that here instead of having to figure out the size of the array by the division of two `sizeof` operations. Might make the code a bit more readable for C++ novices.